### DISCUSSION TITLE
The best overload for 'ChatRequest' does not have a parameter named 'functions'

### DISCUSSION THREAD

**Discussion by Velocedge** [RageAgainstThePixel/OpenAI-DotNet#291] [Q&A] [answered]
# Bug Report

## Overview
I have not updated the NuGet package for Core 7.0  in a while and when I did from 7.3.3 to 7.7.7, I got the error in the title when calling ChatRequest:
```csharp
                var functions = new List<Function>
                {
                new Function(
                    "CategoryIdentification",
                    "Identify a specific category that best described the type of question that was just created from text.",
                     new JsonObject
                     {
                         ["type"] = "object",
                         ["properties"] = new JsonObject
                         {
                            ["categoryText"] = new JsonObject
                            {
                                ["type"] = "string",
                                ["description"] = "The category description itself."
                            }
                         },
                         ["required"] = new JsonArray { "categoryText" }
                     })
                };

    var chatRequest = new ChatRequest(messages, functions: functions, functionCall: "auto", model: modelString );
```
ChatRequest used to have the "functions" parameter.  What was it replaced with?

## To Reproduce
1. Create the function
2. Write the ChatRequest call
3. See the error: The best overload for 'ChatRequest' does not have a parameter named 'functions'

## Expected behavior
Expected ChatRequest to have the functions parameter or something like it.  I can't find any mention of how to deal with functions in the new documentation

**Comment by StephenHodgson** [comment, marked as answer]
Functions were deprecated by OpenAI in favor of tools.

You can simply wrap your functions into a new list of Tools:
```csharp
                var tools = new List<Tool>
                {
                new Function(
                    "CategoryIdentification",
                    "Identify a specific category that best described the type of question that was just created from text.",
                     new JsonObject
                     {
                         ["type"] = "object",
                         ["properties"] = new JsonObject
                         {
                            ["categoryText"] = new JsonObject
                            {
                                ["type"] = "string",
                                ["description"] = "The category description itself."
                            }
                         },
                         ["required"] = new JsonArray { "categoryText" }
                     })
                };

    var chatRequest = new ChatRequest(messages, tools: tools, toolChoice: "auto", model: modelString);
```

**Reply by ADefWebserver** [reply]
Also see:
https://github.com/SyncfusionSuccinctlyE-Books/OpenAI-Succinctly/blob/main/Pages/Functions.razor

**Reply by Velocedge** [reply]
That worked great!  I had to fiddle with the result object a bit to get the data but it was there.

Odd thing though... It ran the first time fine.  However, on the second time when I define the "List<Tool> tool = new List<Tool>" I got an error that says The function 'CategoryIdentification' has already ready been registered.  This is running in a Windows Core Service.  If I stop the service and restart, it works the first time again so I guess I need to find out how to "unregister" it and I all will be good.  If there is any documentation that would shed some light on what's going on, I'd appreciate it if you could point me there.

**Reply by StephenHodgson** [reply]
Functions can only be registered once. So you may need to define this list in a static field/property, then pass that as list of tools. 

I had some issues in the past with the functions being overwritten by others of the same name.

There is a cache mechanism under the hood since it is using reflection. So this is to speed it up a bit
